File: android/src/main/java/com/moengage/flutter/MoEngageFlutterPlugin.java
package com.moengage.flutter;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import androidx.annotation.NonNull;
import com.moe.pushlibrary.MoEHelper;
import com.moengage.core.Logger;
import com.moengage.inapp.MoEInAppHelper;
import com.moengage.plugin.base.PluginHelper;
import com.moengage.plugin.base.model.PushService;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Map.Entry;

import static com.moengage.flutter.ConstantsKt.METHOD_NAME_ENABLE_SDK_LOGS;
import static com.moengage.flutter.ConstantsKt.METHOD_NAME_INITIALISE;
import static com.moengage.flutter.ConstantsKt.METHOD_NAME_LOGOUT;
import static com.moengage.flutter.ConstantsKt.METHOD_NAME_OPT_OUT_TRACKING;
import static com.moengage.flutter.ConstantsKt.METHOD_NAME_PUSH_PAYLOAD;
import static com.moengage.flutter.ConstantsKt.METHOD_NAME_PUSH_TOKEN;
import static com.moengage.flutter.ConstantsKt.METHOD_NAME_RESET_APP_CONTEXT;
import static com.moengage.flutter.ConstantsKt.METHOD_NAME_SELF_HANDLED_INAPP;
import static com.moengage.flutter.ConstantsKt.METHOD_NAME_SET_ALIAS;
import static com.moengage.flutter.ConstantsKt.METHOD_NAME_SET_APP_CONTEXT;
import static com.moengage.flutter.ConstantsKt.METHOD_NAME_SET_APP_STATUS;
import static com.moengage.flutter.ConstantsKt.METHOD_NAME_SET_USER_ATTRIBUTE;
import static com.moengage.flutter.ConstantsKt.METHOD_NAME_SET_USER_ATTRIBUTE_LOCATION;
import static com.moengage.flutter.ConstantsKt.METHOD_NAME_SET_USER_ATTRIBUTE_TIMESTAMP;
import static com.moengage.flutter.ConstantsKt.METHOD_NAME_SHOW_IN_APP;
import static com.moengage.flutter.ConstantsKt.METHOD_NAME_TRACK_EVENT;

public class MoEngageFlutterPlugin implements FlutterPlugin, MethodCallHandler {

  private static final String TAG = "MoEngageFlutterPlugin";
  private static Context context;
  private static final Map<String,String> messageQueue =
      Collections.synchronizedMap(new LinkedHashMap<String, String>());
  private static boolean isInitialised;
  private static MethodChannel channel;
  private static final PluginHelper pluginHelper = new PluginHelper();

  public static void registerWith(Registrar registrar) {
    Logger.v(TAG + " registerWith() : Registering MoEngageFlutterPlugin");
    context = registrar.context();
    initPlugin(registrar.messenger());
  }

  @Override public void onAttachedToEngine(@NonNull FlutterPluginBinding binding) {
    Logger.v(TAG + " onAttachedToEngine() : Registering MoEngageFlutterPlugin");
    context = binding.getApplicationContext();
    initPlugin(binding.getBinaryMessenger());
  }

  @Override public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    Logger.v(TAG + " onDetachedFromEngine() : Registering MoEngageFlutterPlugin");
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    try {
      if (call == null) {
        Logger.e(TAG + " onMethodCall() : MethodCall instance is null cannot proceed "
            + "further.");
        return;
      }
      if (context == null) {
        Logger.e(TAG + " onMethodCall() : Context is null cannot proceed further.");
        return;
      }

      Logger.v(TAG + " onMethodCall() : Method " + call.method);

      switch (call.method) {
        case METHOD_NAME_INITIALISE:
          onInitialised();
          pluginHelper.initialize();
          break;
        case METHOD_NAME_ENABLE_SDK_LOGS:
          enableSDKLogs();
          break;
        case METHOD_NAME_SET_USER_ATTRIBUTE:
          setUserAttribute(call);
          break;
        case METHOD_NAME_SET_USER_ATTRIBUTE_LOCATION:
          setUserLocation(call);
          break;
        case METHOD_NAME_TRACK_EVENT:
          trackEvent(call);
          break;
        case METHOD_NAME_SHOW_IN_APP:
          MoEInAppHelper.getInstance().showInApp(context);
          break;
        case METHOD_NAME_LOGOUT:
          MoEHelper.getInstance(context).logoutUser();
          break;
        case METHOD_NAME_SET_ALIAS:
          setAlias(call);
          break;
        case METHOD_NAME_SET_APP_STATUS:
          setAppStatus(call);
          break;
        case METHOD_NAME_SET_USER_ATTRIBUTE_TIMESTAMP:
          setTimestamp(call);
          break;
        case METHOD_NAME_SELF_HANDLED_INAPP:
          getSelfHandledInAPP(call);
          break;
        case METHOD_NAME_SET_APP_CONTEXT:
          setAppContext(call);
          break;
        case METHOD_NAME_RESET_APP_CONTEXT:
          resetAppContext();
          break;
        case METHOD_NAME_PUSH_PAYLOAD:
          passPushPayload(call);
          break;
        case METHOD_NAME_PUSH_TOKEN:
          passPushToken(call);
          break;
        case METHOD_NAME_OPT_OUT_TRACKING:
          optOutTracking(call);

        default:
          Logger.e(TAG + " onMethodCall() : No mapping for this method.");
      }
    } catch (Exception e) {
      Logger.e(TAG + " onMethodCall() : ", e);
    }
  }

  private static void initPlugin(BinaryMessenger binaryMessenger) {
    channel = new MethodChannel(binaryMessenger, "flutter_moengage_plugin");
    MoEngageFlutterPlugin plugin = new MoEngageFlutterPlugin();
    channel.setMethodCallHandler(plugin);
    pluginHelper.setEventCallback(new EventEmitterImpl());
  }

  private void onInitialised() {
    Logger.v(TAG + " onInitialised() : MoEngage Flutter plugin initialised.");
    Logger.v(TAG + " onInitialised() : Message queue: " + messageQueue);
    isInitialised = true;
    synchronized (messageQueue) {
      // Handle all the messages received before the Dart isolate was
      // initialized, then clear the queue.
      for (Entry<String, String> entry : messageQueue.entrySet()) {
        sendCallback(entry.getKey(), entry.getValue());
      }
      messageQueue.clear();
    }
  }

  private void enableSDKLogs() {
    pluginHelper.enableSDKLogs();
  }

  private void setUserAttribute(MethodCall methodCall) {
    try {
      if (methodCall.arguments == null) return;

      String payload =  methodCall.arguments.toString();
      Logger.v(TAG + " setUserAttribute() : Arguments: " + payload);

      pluginHelper.setUserAttribute(context, payload);
    } catch (Exception e) {
      Logger.e(TAG + " setUserAttribute() : exception: ", e);
    }
  }

  private void setUserLocation(MethodCall methodCall) {
    try {
      if (methodCall.arguments == null) return;

      String payload =  methodCall.arguments.toString();
      Logger.v(TAG + " setUserLocation() : Argument: " + payload);

      pluginHelper.setUserAttribute(context, payload);
    } catch (Exception e) {
      Logger.e(TAG + " setUserLocation() : exception: ", e);
    }
  }

  private void trackEvent(MethodCall methodCall) {
    try {
      if (methodCall.arguments == null) {
        Logger.e(TAG + " trackEvent() : Arguments are null, cannot trackEvent");
        return;
      }
      String payload =  (String) methodCall.arguments;
      Logger.v(TAG + " trackEvent() : Argument :" + payload);

      pluginHelper.trackEvent(context, payload);
    } catch (Exception e) {
      Logger.e(TAG + " trackEvent() : exception: ", e);
    }
  }

  private void setAlias(MethodCall methodCall) {
    try {
      if (methodCall.arguments == null) return;

      String payload =  methodCall.arguments.toString();
      Logger.v(TAG + " setAlias() : Argument :" + payload);

      pluginHelper.setAlias(context, payload);
    } catch (Exception e) {
      Logger.e(TAG + " setAlias() : exception: ", e);
    }
  }

  private void setAppStatus(MethodCall methodCall) {
    try {
      if (methodCall.arguments == null) return;

      String payload =  methodCall.arguments.toString();
      Logger.v(TAG + " setAppStatus() : Argument :" + payload);

      pluginHelper.setAppStatus(context, payload);
    } catch (Exception e) {
      Logger.e(TAG + " setAppStatus() : exception: ", e);
    }
  }

  private void setTimestamp(MethodCall methodCall) {
    try {
      if (methodCall.arguments == null) return;

      String payload =  methodCall.arguments.toString();
      Logger.v(TAG + " setTimestamp() : Arguments: " + payload);

      pluginHelper.setUserAttribute(context, payload);
    } catch (Exception e) {
      Logger.e(TAG + " setTimestamp() : exception: ", e);
    }
  }

  private void getSelfHandledInAPP(MethodCall methodCall) {
    pluginHelper.getSelfHandledInApp(context);
  }

  private void setAppContext(MethodCall methodCall) {
    try {
      if (methodCall.arguments == null) return;

      String payload =  methodCall.arguments.toString();
      Logger.v(TAG + " setAppContext() : Arguments: " + payload);

      pluginHelper.setAppContext(context, payload);
    } catch (Exception e) {
      Logger.e(TAG + " setAppContext() : exception: ", e);
    }
  }

  private void resetAppContext() {
    pluginHelper.resetAppContext(context);
  }

  private void passPushToken(MethodCall methodCall) {
    try {
      if (methodCall.arguments == null) return;

      String payload =  methodCall.arguments.toString();
      Logger.v(TAG + " passPushToken() : Arguments: " + payload);

      pluginHelper.passPushToken(context, payload, PushService.FCM);
    } catch (Exception e) {
      Logger.e(TAG + " passPushToken() : exception: ", e);
    }
  }

  private void passPushPayload(MethodCall methodCall) {
    try {
      if (methodCall.arguments == null) return;

      String payload =  methodCall.arguments.toString();
      Logger.v(TAG + " passPushPayload() : Arguments: " + payload);

      pluginHelper.passPushPayload(context, payload,
          PushService.FCM);
    } catch (Exception e) {
      Logger.e(TAG + " passPushPayload() : exception: ", e);
    }
  }

  private void optOutTracking(MethodCall methodCall) {
    try {
      if (methodCall.arguments == null) return;

      String payload =  methodCall.arguments.toString();
      Logger.v(TAG + " optOutTracking() : Arguments: " + payload);

      pluginHelper.optOutTracking(context, payload);
    } catch (Exception e) {
      Logger.e(TAG + " optOutTracking() : exception: ", e);
    }
  }

  static void sendOrQueueCallback(String methodName, String message) {
    if (isInitialised) {
      Logger.v(TAG + " sendOrQueueCallback() : Flutter Engine initialised will send message");
      sendCallback(methodName, message);
    } else {
      Logger.v(TAG + " sendOrQueueCallback() : Flutter Engine not initialised adding message to "
          + "queue");
      messageQueue.put(methodName, message);
    }
  }

  private static void sendCallback(final String methodName, final String message) {
    new Handler(Looper.getMainLooper()).post(new Runnable() {
      @Override public void run() {
        channel.invokeMethod(methodName, message);
      }
    });
  }

}

File: core/android/src/main/kotlin/com/moengage/flutter/MoEngageFlutterPlugin.kt
package com.moengage.flutter

import android.content.Context
import android.os.Handler
import android.os.Looper
import com.moengage.core.LogLevel
import com.moengage.core.internal.logger.Logger
import com.moengage.plugin.base.internal.PluginHelper
import com.moengage.plugin.base.internal.setEventEmitter
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.FlutterPlugin.FlutterPluginBinding
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler

class MoEngageFlutterPlugin : FlutterPlugin, MethodCallHandler {

    private val tag = "${MODULE_TAG}MoEngageFlutterPlugin"
    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private val pluginHelper = PluginHelper()

    override fun onAttachedToEngine(binding: FlutterPluginBinding) {
        Logger.print { "$tag onAttachedToEngine() : Registering MoEngageFlutterPlugin" }
        context = binding.applicationContext
        initPlugin(binding.binaryMessenger)
    }

    override fun onDetachedFromEngine(binding: FlutterPluginBinding) {
        try {
            Logger.print { "$tag onDetachedFromEngine() : Registering MoEngageFlutterPlugin" }
            //TODO Add API in base plugin if required
//            pluginHelper.onFrameworkDetached()
            channel.setMethodCallHandler(null)
        } catch (t: Throwable) {
            Logger.print(LogLevel.ERROR, t) { "$tag onDetachedFromEngine() " }
        }
    }

    private fun initPlugin(binaryMessenger: BinaryMessenger) {
        try {
            channel = MethodChannel(binaryMessenger, FLUTTER_PLUGIN_CHANNEL_NAME)
            channel.setMethodCallHandler(this)
            setEventEmitter(EventEmitterImpl(::sendCallback))
        } catch (t: Throwable) {
            Logger.print(LogLevel.ERROR, t) { "$tag initPlugin() : exception : " }
        }
    }

    private fun sendCallback(methodName: String, message: String) {
        try {
            Handler(Looper.getMainLooper()).post {
                try {
                    channel.invokeMethod(methodName, message)
                } catch (t: Throwable) {
                    Logger.print(LogLevel.ERROR, t) { "$tag sendCallback() " }
                }
            }
        } catch (t: Throwable) {
            Logger.print(LogLevel.ERROR, t) { "$tag sendCallback() : " }
        }
    }

    @Suppress("SENSELESS_COMPARISON")
    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        try {
            if (call == null) {
                Logger.print(LogLevel.ERROR) { "$tag onMethodCall() : MethodCall instance is null cannot proceed further." }
                return
            }
            if (context == null) {
                Logger.print(LogLevel.ERROR) {
                    "$tag onMethodCall() : Context is null cannot " +
                            "proceed further."
                }
                return
            }
            Logger.print { "$tag onMethodCall() : Method ${call.method}" }
            when (call.method) {
                METHOD_NAME_INITIALISE -> onInitialised(call)
                METHOD_NAME_SET_USER_ATTRIBUTE -> setUserAttribute(call)
                METHOD_NAME_SET_USER_ATTRIBUTE_LOCATION -> setUserLocation(call)
                METHOD_NAME_TRACK_EVENT -> trackEvent(call)
                METHOD_NAME_SHOW_IN_APP -> showInApp(call)
                METHOD_NAME_LOGOUT -> logout(call)
                METHOD_NAME_SET_ALIAS -> setAlias(call)
                METHOD_NAME_SET_APP_STATUS -> setAppStatus(call)
                METHOD_NAME_SET_USER_ATTRIBUTE_TIMESTAMP -> setTimestamp(call)
                METHOD_NAME_SELF_HANDLED_INAPP -> getSelfHandledInApp(call)
                METHOD_NAME_SET_APP_CONTEXT -> setAppContext(call)
                METHOD_NAME_RESET_APP_CONTEXT -> resetAppContext(call)
                METHOD_NAME_PUSH_PAYLOAD -> passPushPayload(call)
                METHOD_NAME_PUSH_TOKEN -> passPushToken(call)
                METHOD_NAME_OPT_OUT_TRACKING -> optOutTracking(call)
                METHOD_NAME_SELF_HANDLED_CALLBACK -> selfHandledCallback(call)
                METHOD_NAME_UPDATE_SDK_STATE -> updateSdkState(call)
                METHOD_NAME_ON_ORIENTATION_CHANGED -> onOrientationChanged()
                else -> Logger.print(LogLevel.ERROR) {
                    "$tag onMethodCall() : No mapping for this" +
                            " method."
                }
            }
        } catch (t: Throwable) {
            Logger.print(LogLevel.ERROR, t) { "$tag onMethodCall() : " }
        }
    }

    private fun logout(methodCall: MethodCall) {
        if (methodCall.arguments == null) return
        val payload = methodCall.arguments.toString()
        Logger.print { "$tag logout() : $payload" }
        pluginHelper.logout(context, payload)
    }

    private fun showInApp(methodCall: MethodCall) {
        if (methodCall.arguments == null) return
        val payload = methodCall.arguments.toString()
        Logger.print { "$tag showInApp() : $payload" }
        pluginHelper.showInApp(context, payload)
    }

    private fun onInitialised(methodCall: MethodCall) {
        if (methodCall.arguments == null) return
        val payload = methodCall.arguments.toString()
        Logger.print { "$tag onInitialised() : MoEngage Flutter plugin initialised." }
        pluginHelper.initialise(payload)
    }

    private fun setUserAttribute(methodCall: MethodCall) {
        try {
            if (methodCall.arguments == null) return
            val payload = methodCall.arguments.toString()
            Logger.print { "$tag setUserAttribute() : Arguments: $payload" }
            pluginHelper.setUserAttribute(context, payload)
        } catch (t: Throwable) {
            Logger.print(LogLevel.ERROR, t) { "$tag setUserAttribute() : " }
        }
    }

    private fun setUserLocation(methodCall: MethodCall) {
        try {
            if (methodCall.arguments == null) return
            val payload = methodCall.arguments.toString()
            Logger.print { "$tag setUserLocation() : Argument: $payload" }
            pluginHelper.setUserAttribute(context, payload)
        } catch (t: Throwable) {
            Logger.print(LogLevel.ERROR, t) { "$tag setUserLocation() : " }
        }
    }

    private fun trackEvent(methodCall: MethodCall) {
        try {
            if (methodCall.arguments == null) {
                Logger.print(LogLevel.ERROR) {
                    "$tag trackEvent() : Arguments are null, cannot" +
                            " trackEvent"
                }
                return
            }
            val payload = methodCall.arguments as String
            Logger.print { "$tag trackEvent() : Argument :$payload" }
            pluginHelper.trackEvent(context, payload)
        } catch (t: Throwable) {
            Logger.print(LogLevel.ERROR, t) { "$tag trackEvent() : " }
        }
    }

    private fun setAlias(methodCall: MethodCall) {
        try {
            if (methodCall.arguments == null) return
            val payload = methodCall.arguments.toString()
            Logger.print { "$tag setAlias() : Argument :$payload" }
            pluginHelper.setAlias(context, payload)
        } catch (t: Throwable) {
            Logger.print(LogLevel.ERROR, t) { "$tag setAlias() : " }
        }
    }

    private fun setAppStatus(methodCall: MethodCall) {
        try {
            if (methodCall.arguments == null) return
            val payload = methodCall.arguments.toString()
            Logger.print { "$tag setAppStatus() : Argument :$payload" }
            pluginHelper.setAppStatus(context, payload)
        } catch (t: Throwable) {
            Logger.print(LogLevel.ERROR, t) { "$tag setAppStatus() : " }
        }
    }

    private fun setTimestamp(methodCall: MethodCall) {
        try {
            if (methodCall.arguments == null) return
            val payload = methodCall.arguments.toString()
            Logger.print { "$tag setTimestamp() : Arguments: $payload" }
            pluginHelper.setUserAttribute(context, payload)
        } catch (t: Throwable) {
            Logger.print(LogLevel.ERROR, t) { "$tag setTimestamp() : " }
        }
    }

    private fun getSelfHandledInApp(methodCall: MethodCall) {
        try {
            if (methodCall.arguments == null) return
            val payload = methodCall.arguments.toString()
            Logger.print { "$tag getSelfHandledInApp() : Arguments: $payload" }
            pluginHelper.getSelfHandledInApp(context, payload)
        } catch (t: Throwable) {
            Logger.print(LogLevel.ERROR, t) { "$tag getSelfHandledInApp() : " }
        }
    }

    private fun setAppContext(methodCall: MethodCall) {
        try {
            if (methodCall.arguments == null) return
            val payload = methodCall.arguments.toString()
            Logger.print { "$tag setAppContext() : Arguments: $payload" }
            pluginHelper.setAppContext(context, payload)
        } catch (t: Throwable) {
            Logger.print(LogLevel.ERROR, t) { "$tag setAppContext() : " }
        }
    }

    private fun resetAppContext(methodCall: MethodCall) {
        try {
            if (methodCall.arguments == null) return
            val payload = methodCall.arguments.toString()
            Logger.print { "$tag resetAppContext() : Arguments: $payload" }
            pluginHelper.resetAppContext(context, payload)
        } catch (t: Throwable) {
            Logger.print(LogLevel.ERROR, t) { "$tag resetAppContext() : " }
        }
    }

    private fun passPushToken(methodCall: MethodCall) {
        try {
            if (methodCall.arguments == null) return
            val payload = methodCall.arguments.toString()
            Logger.print { "$tag passPushToken() : Arguments: $payload" }
            pluginHelper.passPushToken(context, payload)
        } catch (t: Throwable) {
            Logger.print(LogLevel.ERROR, t) { "$tag passPushToken() : " }
        }
    }

    private fun passPushPayload(methodCall: MethodCall) {
        try {
            if (methodCall.arguments == null) return
            val payload = methodCall.arguments.toString()
            Logger.print { "$tag passPushPayload() : Arguments: $payload" }
            pluginHelper.passPushPayload(context, payload)
        } catch (t: Throwable) {
            Logger.print(LogLevel.ERROR, t) { "$tag passPushPayload() : " }
        }
    }

    private fun optOutTracking(methodCall: MethodCall) {
        try {
            if (methodCall.arguments == null) return
            val payload = methodCall.arguments.toString()
            Logger.print { "$tag optOutTracking() : Arguments: $payload" }
            pluginHelper.optOutTracking(context, payload)
        } catch (t: Throwable) {
            Logger.print(LogLevel.ERROR, t) { "$tag optOutTracking() : " }
        }
    }

    private fun selfHandledCallback(methodCall: MethodCall) {
        try {
            if (methodCall.arguments == null) return
            val payload = methodCall.arguments.toString()
            Logger.print { "$tag selfHandledCallback() : Arguments: $payload" }
            pluginHelper.selfHandledCallback(context, payload)
        } catch (t: Throwable) {
            Logger.print(LogLevel.ERROR, t) { "$tag selfHandledCallback() : " }
        }
    }

    private fun updateSdkState(methodCall: MethodCall) {
        try {
            if (methodCall.arguments == null) return
            val payload = methodCall.arguments.toString()
            Logger.print { "$tag selfHandledCallback() : Arguments: $payload" }
            pluginHelper.storeFeatureStatus(context, payload)
        } catch (t: Throwable) {
            Logger.print(LogLevel.ERROR, t) { "$tag selfHandledCallback() : " }
        }
    }

    private fun onOrientationChanged() {
        Logger.print { "$tag onOrientationChanged() : " }
        pluginHelper.onConfigurationChanged()
    }
}